streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/SuppressImpl.java

```diff
@@ -0,0 +1,166 @@
+            return numberOfKeysToRemember;
+        }
+
+        long getBytesToUseForSuppressionStorage() {
+            return bytesToUseForSuppressionStorage;
+        }
+
+        public BufferFullStrategy getBufferFullStrategy() {
+            return bufferFullStrategy;
+        }
+
+        public Serializer<K> getKeySerializer() {
+            return keySerializer;
+        }
+
+        public Serializer<V> getValueSerializer() {
+            return valueSerializer;
+        }
+    }
+
+    public static class IntermediateSuppressionImpl<K, V> implements Suppress.IntermediateSuppression<K, V> {
+        private BufferConfigImpl<K, V> bufferConfig = new BufferConfigImpl<>();
+        private Duration timeToWaitForMoreEvents = null;
+
+        public IntermediateSuppressionImpl() {}
+
+        private IntermediateSuppressionImpl(final IntermediateSuppressionImpl<K, V> from) {
+            timeToWaitForMoreEvents = from.timeToWaitForMoreEvents;
+            this.bufferConfig = from.bufferConfig;
+        }
+
+        @Override
+        public IntermediateSuppression<K, V> emitAfter(final Duration timeToWaitForMoreEvents) {
+            final IntermediateSuppressionImpl<K, V> result = new IntermediateSuppressionImpl<>(this);
+            result.timeToWaitForMoreEvents = timeToWaitForMoreEvents;
+            return result;
+        }
+
+        @Override
+        public IntermediateSuppression<K, V> bufferConfig(final BufferConfig<K, V> bufferConfig) {
+            final IntermediateSuppressionImpl<K, V> result = new IntermediateSuppressionImpl<>(this);
+            result.bufferConfig = (BufferConfigImpl<K, V>) bufferConfig;
+            return result;
+        }
+
+        Duration getTimeToWaitForMoreEvents() {
+            return timeToWaitForMoreEvents;
+        }
+
+        BufferConfigImpl<K, V> getBufferConfig() {
+            return bufferConfig;
+        }
+    }
+
+    private SuppressImpl() {}
+
+    public SuppressImpl(final BufferConfig<K, V> finalResultsConfig) {
+        this.finalResultsConfig = finalResultsConfig;
+    }
+
+    public SuppressImpl(final IntermediateSuppression<K, V> intermediateSuppression) {
+        this.intermediateSuppression = (IntermediateSuppressionImpl<K, V>) intermediateSuppression;
+    }
+
+    IntermediateSuppressionImpl<K, V> getIntermediateSuppression() {
+        return intermediateSuppression;
+    }
+
+    TimeDefinition<K, V> getTimeDefinition() {
+        return timeDefinition;
+    }
+
+    public boolean isFinalResultsSuppression() {
+        return finalResultsConfig != null;
+    }
+
+    boolean isImmediateEmit() {
+        return intermediateSuppression == null
+            || intermediateSuppression.getTimeToWaitForMoreEvents() != null
+            && intermediateSuppression.getTimeToWaitForMoreEvents().toMillis() <= 0L;
+    }
+
+    public static <K extends Windowed, V> Suppress<K, V> buildFinalResultsSuppression(final Duration windowCloseTime,
```

Comment:
`Suppress` is split into interface/impl mostly to support these kinds of internal methods.